File: python/799_Champagne_Tower/solution.py
class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        tower = [[0] * (r+1) for r in range(query_row + 1)]
        tower[0][0] = poured
        for row in range(len(tower)):
            for g, vol in enumerate(tower[row]):
                if vol > 1:
                    if row != len(tower) - 1:
                        excess = (vol - 1) / 2
                        tower[row+1][g] += excess
                        tower[row+1][g+1] += excess
                    tower[row][g] = 1
        return tower[query_row][query_glass]

File: rust/799_Champagne_Tower/solution.rs
impl Solution {
    pub fn champagne_tower(poured: i32, query_row: i32, query_glass: i32) -> f64 {
        let mut tower: Vec<Vec<f64>> = vec![vec![poured as f64]];
        for row in 1..=(query_row) {
            let mut next_row: Vec<f64> = vec![0.0; (row + 1) as usize];
            for glass in 0..tower.len() {
                let volume = tower[(row - 1) as usize][glass];
                if volume > 1.0 {
                    let runoff = (volume - 1.0) / 2.0;
                    next_row[glass] += runoff;
                    next_row[glass+1] += runoff;
                }
            }
            tower.push(next_row)
        }
        if tower[query_row as usize][query_glass as usize] > 1.0 {
            1.0
        } else {
            tower[query_row as usize][query_glass as usize]
        }
    }
}
